File: app/src/main/java/moe/tlaster/openween/common/helpers/HttpHelper.java
package moe.tlaster.openween.common.helpers;

import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.zhy.http.okhttp.OkHttpUtils;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import moe.tlaster.openween.core.api.Entity;

/**
 * Created by Tlaster on 2016/9/5.
 */
public class HttpHelper {
    public static void getAsync(String url, Map<String, String> param, @NonNull JsonCallback callback) {
        if (TextUtils.isEmpty(Entity.getAccessToken())){
            callback.onError(null, new InvalidAccessTokenException(), -1);
            return;
        }
        getAsync(url, Entity.getAccessToken(), param, callback);
    }

    public static void getAsync(String url, String token, Map<String, String> param, @NonNull JsonCallback callback){
        if (param == null) param = new HashMap<>();
        param.put("access_token", token);
        OkHttpUtils.get().url(url).params(param).build().execute(callback);
    }

    public static void postAsync(String url, Map<String, String> data, @NonNull JsonCallback callback) {
        if (TextUtils.isEmpty(Entity.getAccessToken())){
            callback.onError(null, new InvalidAccessTokenException(), -1);
            return;
        }
        if (data == null) data = new HashMap<>();
        data.put("access_token", Entity.getAccessToken());
        OkHttpUtils.post().url(url).params(data).build().execute(callback);
    }

    public static void uploadFileAsync(String url, File file, @NonNull JsonCallback callback) {
        if (TextUtils.isEmpty(Entity.getAccessToken())){
            callback.onError(null, new InvalidAccessTokenException(), -1);
            return;
        }
        OkHttpUtils.post().url(url).addFile("pic","pic.png",file).addParams("access_token", Entity.getAccessToken()).build().execute(callback);
    }

    public static void uploadFileWithParamAsync(String url, File file, Map<String, String> data, @NonNull JsonCallback callback) {
        if (TextUtils.isEmpty(Entity.getAccessToken())){
            callback.onError(null, new InvalidAccessTokenException(), -1);
            return;
        }
        if (data == null) data = new HashMap<>();
        data.put("access_token", Entity.getAccessToken());
        OkHttpUtils.post().url(url).addFile("pic","pic.png",file).params(data).build().execute(callback);
    }
}

File: app/src/main/java/moe/tlaster/openween/common/helpers/HttpHelper.kt
package moe.tlaster.openween.common.helpers

import android.support.v4.util.ArrayMap
import android.text.TextUtils
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.github.kittinunf.fuel.android.core.Json
import com.github.kittinunf.fuel.android.extension.responseJson
import com.github.kittinunf.fuel.core.FuelError
import com.github.kittinunf.fuel.core.HttpException
import com.github.kittinunf.fuel.core.Request
import com.github.kittinunf.fuel.core.Response
import com.github.kittinunf.fuel.httpGet
import com.github.kittinunf.fuel.httpPost
import com.github.kittinunf.fuel.httpUpload
import com.github.kittinunf.result.Result
import moe.tlaster.openween.common.WeiboException
import java.io.File

/**
 * Created by Tlaster on 2017/1/19.
 */
object HttpHelper {
    inline fun  <reified T: Any> getAsync(url: String, param: Map<String, Any>) : T {
        (param as ArrayMap).put("access_token", Setttings.AccessToken)
        if (T::class.java == String::class.java) {
            return checkForError(url.httpGet(param.toList()).responseJson().third) as T
        } else {
            return ObjectMapper().readValue(checkForError(url.httpGet(param.toList()).responseJson().third))
        }
    }

    //    fun getAsync(url: String, param: Map<String, Any?>?) : String {
//        return checkForError(url.httpGet(param?.toList()).responseJson().third)
//    }
    inline fun  <reified T: Any> postAsync(url: String, param: Map<String, Any>, file: File? = null) : T {
        (param as ArrayMap).put("access_token", Setttings.AccessToken)
        val request = if (file == null) {
            url.httpPost(param.toList())
        } else {
            url.httpUpload(parameters = param.toList()).source { _, _ -> file }
        }
        if (T::class.java == String::class.java) {
            return checkForError(request.responseJson().third) as T
        } else {
            return ObjectMapper().readValue(checkForError(request.responseJson().third))
        }
    }

    fun checkForError(responseJson: Result<Json, FuelError>): String {
        if (responseJson.component2()?.exception != null) throw responseJson.component2()!!
        val value = responseJson.component1()!!
        if (!TextUtils.isEmpty(value.obj().optString("errorCode"))) throw WeiboException(value.obj().optString("error"))
        return value.content
    }
}